source/MRMesh/MRMeshComponents.cpp

```diff
@@ -468,6 +499,99 @@ UnionFind<FaceId> getUnionFindStructureFacesPerEdge( const MeshPart& meshPart, c
 
     return res;
 }
+
+UnionFind<FaceId> getUnionFindStructureFacesPerEdgeWithZTolerance( const MeshPart& meshPart, const float zTolerance )
+{
+    MR_TIMER
+
+    const auto& mesh = meshPart.mesh;
+    const FaceBitSet& region = mesh.topology.getFaceIds( meshPart.region );
+    const FaceBitSet* lastPassFaces = &region;
+    const auto numFaces = region.find_last() + 1;
+    UnionFind<FaceId> res( numFaces );
+    const auto numThreads = int( tbb::global_control::active_value( tbb::global_control::max_allowed_parallelism ) );
+    FaceBitSet bdFaces;
+
+    Vector<Vector2f, FaceId> zRanges( numFaces );
+    tbb::parallel_for( tbb::blocked_range<FaceId>( FaceId( 0 ), FaceId( numFaces ) ),
+            [&] ( const tbb::blocked_range<FaceId>& range )
+    {
+        for ( auto f = range.begin(); f < range.end(); ++f )
+        {
+            ThreeVertIds ids;
+            meshPart.mesh.topology.getTriVerts( f, ids );
+            zRanges[f].x = std::min( { meshPart.mesh.points[ids[0]].z, meshPart.mesh.points[ids[1]].z, meshPart.mesh.points[ids[2]].z } );
+            zRanges[f].y = std::max( { meshPart.mesh.points[ids[0]].z, meshPart.mesh.points[ids[1]].z, meshPart.mesh.points[ids[2]].z } );
+        }
+    } );
+
+    if ( numThreads > 1 )
+    {
+        bdFaces.resize( numFaces );
+        lastPassFaces = &bdFaces;
+        const int endBlock = int( bdFaces.size() + bdFaces.bits_per_block - 1 ) / bdFaces.bits_per_block;
+        const auto bitsPerThread = ( endBlock + numThreads - 1 ) / numThreads * BitSet::bits_per_block;
+
+        std::shared_mutex mutex;
+
+        tbb::parallel_for( tbb::blocked_range<int>( 0, numThreads ),
+            [&] ( const tbb::blocked_range<int>& range )
+        {
+            const FaceId fBeg{ range.begin() * bitsPerThread };
+            const FaceId fEnd{ range.end() < numThreads ? range.end() * bitsPerThread : bdFaces.size() };
+            for ( auto f0 = fBeg; f0 < fEnd; ++f0 )
+            {
+                if ( !contains( region, f0 ) )
+                    continue;
+                EdgeId e[3];
+                mesh.topology.getTriEdges( f0, e );
+                for ( int i = 0; i < 3; ++i )
+                {
+                    assert( mesh.topology.left( e[i] ) == f0 );
+                    FaceId f1 = mesh.topology.right( e[i] );
+                    if ( f0 < f1 && contains( meshPart.region, f1 ) )
+                    {
+                        const float zMax = std::max( zRanges[f0].y, zRanges[f1].y );
+                        const float zMin = std::min( zRanges[f0].x, zRanges[f1].x );
+                        
+                        if ( f1 < fEnd && zMax - zMin < zTolerance )
+                        {
+                            res.unite( f0, f1 ); // our region
+                            std::shared_lock<std::shared_mutex> lock( mutex );
```

Comment:
why shared lock and not simple lock? (also why shared_mutex here?)

Reply:
if we use mutex why not realize it with 
```c++
[[nodiscard]] MRMESH_API std::vector<FaceBitSet> getAllComponents( const MeshPart& meshPart,
    FaceIncidence incidence = FaceIncidence::PerEdge, const UndirectedEdgePredicate & isCompBd = {} );
```
just hide all the code that managed the map inside?